virtool/users/sessions.py

```diff
@@ -2,219 +2,220 @@
 import hashlib
 import json
 import secrets
+from datetime import timedelta
 from typing import Optional, Tuple
 
-import arrow
 from aioredis import Redis
-from motor.motor_asyncio import AsyncIOMotorClientSession
+from virtool_core.models.session import (
+    Session,
+    SessionAuthentication,
+)
 
 import virtool.utils
-from virtool.api.custom_json import dumps
-from virtool.mongo.core import DB
-from virtool.types import Document
-
-
-async def create_session(
-    db: DB,
-    redis: Redis,
-    ip: str,
-    user_id: Optional[str] = None,
-    remember: Optional[bool] = False,
-    session: Optional[AsyncIOMotorClientSession] = None,
-) -> Tuple[str, Document, str]:
-    session_id = await create_session_id(redis)
-
-    utc = arrow.utcnow()
-
-    if user_id and remember:
-        expires_at = utc.shift(days=30)
-    elif user_id:
-        expires_at = utc.shift(minutes=60)
-    else:
-        expires_at = utc.shift(minutes=10)
-
-    new_session = {
-        "created_at": virtool.utils.timestamp().timestamp(),
-        "ip": ip,
-    }
-
-    token = None
-
-    if user_id:
-        token, hashed = virtool.utils.generate_key()
-        user_document = await db.users.find_one(user_id, session=session)
-        new_session.update(
-            {
-                "token": hashed,
-                "administrator": user_document["administrator"],
-                "groups": user_document["groups"],
-                "permissions": user_document["permissions"],
-                "force_reset": user_document["force_reset"],
-                "user": {"id": user_id},
-            }
-        )
-
-    await redis.set(session_id, dumps(new_session))
-    await redis.expireat(session_id, expires_at.timestamp())
-
-    return session_id, new_session, token
-
-
-async def create_session_id(redis: Redis) -> str:
-    """
-    Create a new unique session id.
+from virtool.api.custom_json import dumps, fromisoformat, isoformat
+from virtool.data.errors import ResourceError
+from virtool.data.piece import DataLayerPiece
+
+
+class SessionData(DataLayerPiece):
+    def __init__(self, redis: Redis):
+        self.redis = redis
+
+    async def create(
```

Comment:
At a minimum, I think we should break this into `create_session` and `create_anonymous_session`.

The method signature bothers me because `remember` is not used for anonymous sessions nor is the authentication branch visited.